dags/sentinel-2_gap_filler.py

```diff
@@ -0,0 +1,200 @@
+            "DataType": "String",
+            "StringValue": maturity,
+        }
+
+    bbox = stac_doc.get("bbox")
+    if bbox and len(bbox) > 3:
+        msg_attributes["bbox.ll_lon"] = {
+            "DataType": "Number",
+            "StringValue": str(bbox[0]),
+        }
+        msg_attributes["bbox.ll_lat"] = {
+            "DataType": "Number",
+            "StringValue": str(bbox[1]),
+        }
+        msg_attributes["bbox.ur_lon"] = {
+            "DataType": "Number",
+            "StringValue": str(bbox[2]),
+        }
+        msg_attributes["bbox.ur_lat"] = {
+            "DataType": "Number",
+            "StringValue": str(bbox[3]),
+        }
+
+    return msg_attributes
+
+
+def get_missing_stac_files(s3_report_path, offset=0, limit=None):
+    """
+    read the gap report
+    """
+
+    hook = S3Hook(aws_conn_id=dag.default_args["africa_conn_id"])
+    bucket_name, key = hook.parse_s3_url(s3_report_path)
+    print(f"Reading the gap report took {s3_report_path} Seconds")
+
+    # ToDo: changing the bucket name was due to a bug. Remove this when new data is available.
+    files = (
+        hook.read_key(key=key, bucket_name=bucket_name)
+        .replace("sentinel-cogs-inventory", f"{SRC_BUCKET_NAME}")
+        .splitlines()
+    )
+
+    for f in files[offset:limit]:
+        yield f
+
+
+def publish_messages(messages):
+    """
+    Publish messages to the data transfer queue
+    param message: list of messages
+    """
+
+    for num, message in enumerate(messages):
+        message["Id"] = str(num)
+    sqs_hook = SQSHook(aws_conn_id=dag.default_args["us_conn_id"])
+    sqs = sqs_hook.get_resource_type("sqs")
+    queue = sqs.get_queue_by_name(QueueName=QUEUE_NAME)
+    queue.send_messages(Entries=messages)
+
+
+def get_contents_and_attributes(hook, s3_filepath):
+    bucket_name, key = hook.parse_s3_url(s3_filepath)
+    contents = hook.read_key(key=key, bucket_name=SRC_BUCKET_NAME)
+    contents_dict = json.loads(contents)
+    attributes = get_common_message_attributes(contents_dict)
+    return contents, attributes
+
+
+def prepare_message(hook, s3_path):
+    """
+    Prepare a single message for each stac file
+    """
+
+    key_exists = hook.check_for_key(s3_path)
+    if not key_exists:
+        raise ValueError(f"{s3_path} does not exist")
+
+    contents, attributes = get_contents_and_attributes(hook, s3_path)
+    message = {
+        "MessageBody": json.dumps(
+            {"Message": contents, "MessageAttributes": attributes}
+        ),
+    }
+    return message
+
+
+def prepare_and_send_messages(dag_run, **kwargs):
+    hook = S3Hook(aws_conn_id=dag.default_args["us_conn_id"])
+    # Read the missing stac files from the gap report file
+    files = get_missing_stac_files(
+        dag_run.conf["s3_report_path"], dag_run.conf["offset"], dag_run.conf["limit"]
+    )
+
+    max_workers = 10
+    # counter for files that no longer exist
+    failed = 0
+
+    batch = []
```

Comment:
Is this thing going to be thread safe, @omad? Is there not a race between appending items and emptying the list?

Reply:
I believe so, we're only modifying the list in the main thread.